In [3]:
import pandas as pd

df = pd.read_csv('../Planilhas/imoveis_unificado.csv', encoding='utf-8', sep=';')

C:\Users\Leo\AppData\Local\Temp\ipykernel_10112\869521274.py:3: DtypeWarning: Columns (0: numero, 1: ano_construcao, 2: cod_logradouro, 3: latitude, 4: longitude, 5: ano, 6: distrito) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../Planilhas/imoveis_unificado.csv', encoding='utf-8', sep=';')


In [9]:
df.info()
df.head(20)


<class 'pandas.DataFrame'>
RangeIndex: 44234 entries, 0 to 44233
Data columns (total 23 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   logradouro          44234 non-null  str   
 1   numero              42788 non-null  object
 2   complemento         41531 non-null  str   
 3   valor_avaliacao     44234 non-null  str   
 4   bairro              44234 non-null  str   
 5   cidade              44234 non-null  str   
 6   uf                  44234 non-null  str   
 7   ano_construcao      44234 non-null  object
 8   area_terreno        44234 non-null  str   
 9   area_construida     44234 non-null  str   
 10  fracao_ideal        44234 non-null  str   
 11  padrao_acabamento   44234 non-null  str   
 12  tipo_construcao     44234 non-null  str   
 13  tipo_ocupacao       44234 non-null  str   
 14  data_transacao      44234 non-null  str   
 15  estado_conservacao  44234 non-null  str   
 16  tipo_imovel         44234 non-nul

,logradouro,numero,complemento,valor_avaliacao,bairro,cidade,uf,ano_construcao,area_terreno,area_construida,...,tipo_ocupacao,data_transacao,estado_conservacao,tipo_imovel,sfh,cod_logradouro,latitude,longitude,ano,distrito
0,logradouro,numero,complemento,valor_avaliacao,bairro,cidade,uf,ano_construcao,area_terreno,area_construida,...,tipo_ocupacao,data_transacao,estado_conservacao,tipo_imovel,sfh,cod_logradouro,latitude,longitude,ano,Distrito
1,av norte miguel arraes de alencar,3071,NaN,1068562.63,Encruzilhada,Recife,PE,1997,"438,0","511,0",...,COMERCIAL COM LIXO ORGANICO,2023-12-21,Regular,Galpão,0.00,46540,-8.034273,-34.896337,2023,1
2,av norte miguel arraes de alencar,3029,NaN,1500000.00,Encruzilhada,Recife,PE,1957,"779,33","582,44",...,COMERCIAL SEM LIXO ORGANICO,2023-11-17,Regular,Casa,0.00,46540,-8.034435,-34.896335,2023,1
3,rua belmiro correa,133,apto 1,110000.00,Encruzilhada,Recife,PE,1970,"562,05","121,0",...,RESIDENCIAL,2023-09-22,Bom,Apartamento,0.00,10715,-8.035013,-34.895903,2023,1
4,rua belmiro correa,133,apto 1,110000.00,Encruzilhada,Recife,PE,1970,"562,05","121,0",...,RESIDENCIAL,2023-09-26,Bom,Apartamento,0.00,10715,-8.035013,-34.895903,2023,1
5,rua belmiro correa,133,apto 2,110000.00,Encruzilhada,Recife,PE,1970,"562,05","81,0",...,RESIDENCIAL,2023-09-22,Bom,Apartamento,0.00,10715,-8.035013,-34.895903,2023,1
6,rua belmiro correa,133,apto 3,110000.00,Encruzilhada,Recife,PE,1970,"562,05","81,0",...,RESIDENCIAL,2023-09-22,Bom,Apartamento,0.00,10715,-8.035013,-34.895903,2023,1
7,rua belmiro correa,133,apto 4,110000.00,Encruzilhada,Recife,PE,1970,"562,05","81,0",...,RESIDENCIAL,2023-09-26,Bom,Apartamento,0.00,10715,-8.035013,-34.895903,2023,1
8,rua belmiro correa,133,apto 0005,110000.00,Encruzilhada,Recife,PE,1970,"562,05","81,0",...,RESIDENCIAL,2023-09-22,Bom,Apartamento,0.00,10715,-8.035013,-34.895903,2023,1
9,rua belmiro correa,109,NaN,4900000.00,Encruzilhada,Recife,PE,1951,"439,28","343,23",...,RESIDENCIAL,2023-09-26,Bom,Casa,0.00,10715,-8.035165,-34.895961,2023,1


###Tratamento das colunas ano_construcao, cod_logradouro, valor_avaliacao###

Começando com a coluna ano_construcao, analisando as informações gerais de qual o estado inicial da coluna

In [23]:
df1 = df['ano_construcao']
df1.info()

print(df['ano_construcao'].value_counts())

print(df['ano_construcao'].describe())

<class 'pandas.Series'>
RangeIndex: 44234 entries, 0 to 44233
Series name: ano_construcao
Non-Null Count  Dtype 
--------------  ----- 
44234 non-null  object
dtypes: object(1)
memory usage: 345.7+ KB
ano_construcao
2025      2192
2023      1296
2024      1248
2024      1121
2025       980
          ... 
670,0        1
1237,5       1
421,25       1
1943         1
1944         1
Name: count, Length: 3263, dtype: int64
count     44234
unique     3263
top        2025
freq       2192
Name: ano_construcao, dtype: int64


Como podemos ver por essas informações, teremos alguns pontos a serem tratados: O tipo de variável está incorreto, faz mais sentido estar como um int, por ser um ano; Já que não temos valores nulos, não precisamos nos preocupar com isso. O principal ponto de atenção são os anos sem coerência, como 670 (que são muitos antigos), ou valores que têm decimal. Felizmente, não tem valores acima de 2025, então nada sobresaiu o teto.

In [ ]:
print(df['ano_construcao'].unique())# Verifica o que está impedindo a conversão para int


df['ano_construcao'] = pd.to_numeric(df['ano_construcao'], errors='coerce').astype('Int64')# Converte forçando os erros para nulo 

print(df['ano_construcao'].dtype)
print(df['ano_construcao'].isna().sum())  # quantos viraram nulo

df['ano_construcao'].info()

<IntegerArray>
[<NA>, 1997, 1957, 1970, 1951, 2017, 2011, 2021, 2002, 2010, 2016, 2015, 1984,
 2014, 2013, 2008, 2012, 2009, 1939, 2003, 1940, 2004, 1995, 1948, 1994, 1988,
 2024, 1979, 1996, 2006, 1959, 2018, 1999, 1964, 1992, 1998, 1993, 1945, 1975,
 2007, 2005, 1986, 1987, 1983, 2001, 2023, 2000, 1941, 1969, 1978, 1980, 1989,
 1966, 2019, 1982, 1954, 1942, 1968, 1985, 1990, 1952, 1977, 2020, 1972, 1991,
 1981, 1955, 1971, 1976, 1973, 1974, 1947, 1946, 1944, 1950, 1956, 2022, 1949,
 1962, 1965, 1943, 1963, 1967, 1961, 1958, 1960, 1953, 2025]
Length: 88, dtype: Int64
Int64
15246
<class 'pandas.Series'>
RangeIndex: 44234 entries, 0 to 44233
Series name: ano_construcao
Non-Null Count  Dtype
--------------  -----
28988 non-null  Int64
dtypes: Int64(1)
memory usage: 388.9 KB


Agora, irei tratar os valores nulos que não foram convertidos corretamente. Para repor esses valores nulos, vou usar a mediana do ano de construção das demais construções que estão no bairro daquela construção, para tornar mais coerente, mantendo as informações sobre bairros mais antigos ou mais novos. Também podemos ver que não tem nenhum valor absurdo como outlier, todas datas estão coerentes para um edifício atual

In [37]:
df['ano_construcao'] = df['ano_construcao'].fillna(
    df.groupby('bairro')['ano_construcao'].transform('median') #comando usado para pegar a mediana dos anos de construção de um determinado bairro, e substituir nos locais nulos
)

df['ano_construcao'].info()
df['cod_logradouro'].head()

<class 'pandas.Series'>
RangeIndex: 44234 entries, 0 to 44233
Series name: ano_construcao
Non-Null Count  Dtype
--------------  -----
44231 non-null  Int64
dtypes: Int64(1)
memory usage: 388.9 KB


0    cod_logradouro
1             46540
2             46540
3             10715
4             10715
Name: cod_logradouro, dtype: object

In [38]:
df.head()

,logradouro,numero,complemento,valor_avaliacao,bairro,cidade,uf,ano_construcao,area_terreno,area_construida,...,tipo_ocupacao,data_transacao,estado_conservacao,tipo_imovel,sfh,cod_logradouro,latitude,longitude,ano,distrito
0,logradouro,numero,complemento,valor_avaliacao,bairro,cidade,uf,<NA>,area_terreno,area_construida,...,tipo_ocupacao,data_transacao,estado_conservacao,tipo_imovel,sfh,cod_logradouro,latitude,longitude,ano,Distrito
1,av norte miguel arraes de alencar,3071,NaN,1068562.63,Encruzilhada,Recife,PE,1997,"438,0","511,0",...,COMERCIAL COM LIXO ORGANICO,2023-12-21,Regular,Galpão,0.00,46540,-8.034273,-34.896337,2023,1
2,av norte miguel arraes de alencar,3029,NaN,1500000.00,Encruzilhada,Recife,PE,1957,"779,33","582,44",...,COMERCIAL SEM LIXO ORGANICO,2023-11-17,Regular,Casa,0.00,46540,-8.034435,-34.896335,2023,1
3,rua belmiro correa,133,apto 1,110000.00,Encruzilhada,Recife,PE,1970,"562,05","121,0",...,RESIDENCIAL,2023-09-22,Bom,Apartamento,0.00,10715,-8.035013,-34.895903,2023,1
4,rua belmiro correa,133,apto 1,110000.00,Encruzilhada,Recife,PE,1970,"562,05","121,0",...,RESIDENCIAL,2023-09-26,Bom,Apartamento,0.00,10715,-8.035013,-34.895903,2023,1
